### PR TITLE
APPS-8 Hide Spryks details

### PR DESCRIPTION
**Pull request by oleksandr-stoilovskyi**
- Developer(s): @oleksandr-stoilovskyi

- Ticket: https://spryker.atlassian.net/browse/APPS-8

- Academy PR: -

- Release Group: -

#### Release Table

   Module                | Release Type         | Constraint Updates         |
   :--------------------- | :------------------------ | :--------------------- |
   Spryk               | minor                 |            |

-----------------------------------------

##### Change log

Improvements

- Adjusted `SprykFacadeInterface::getSprykDefinitions()` so now it accepts optional parameter level which filters the result instead of no optional parameters and no filtering by level.
- Impacted `SprykDumpConsole` with facade changes.

### CODE REVIEW

File: src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+ * Copyright © 2016-present Spryker Systems GmbH. All rights reserved.
+ * Use of this software requires acceptance of the Evaluation License Agreement. See LICENSE file.
+ */
+
+namespace SprykerSdk\Spryk\Model\Spryk\Configuration\Validator\Rules;
+
+class LevelRule implements ConfigurationValidatorRuleInterface
+{
+    protected const LEVEL_CONFIG_NAME = 'level';
+    /**
```
Comment (pushokwhite) [inline src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php:13]: Please add empty string after the constant.

File: src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+     * @var string
+     */
+    protected $errorMessage = 'Level is required value of configuration';
+
+    /**
+     * @var array
```
Comment (pushokwhite) [inline src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php:19]:
```suggestion
     * @var int[]
```

File: src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+     */
+    protected $availableLevels;
+
+    public function __construct(array $availableLevels)
```
Comment (pushokwhite) [inline src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php:23]: PHPDoc is missing.

File: src/SprykerSdk/Spryk/SprykConfig.php [hunk 4]
```diff
@@ -138,6 +154,11 @@ public function getDefaultDevelopmentMode(): string
         return static::NAME_DEVELOPMENT_LAYER_PROJECT;
     }
 
+    public function getDefaultLevel(): int
```
Comment (pushokwhite) [inline src/SprykerSdk/Spryk/SprykConfig.php:157]: Too ambiguous. You have default value in console command and config loader and they are different from each other.

File: config/spryk/spryks/Spryker/AddGuiModule.yml [hunk 5]
```diff
@@ -2,6 +2,7 @@ spryk: structure
 description: "Adds a new GUI *module* structure."
 priority: 1
 mode: both
+level: 1
```
[the remark this answers is not in the document]
Comment (oleksandr-stoilovskyi) [reply config/spryk/spryks/Spryker/AddGuiModule.yml:5]: added

File: src/SprykerSdk/Spryk/Console/SprykDumpConsole.php [hunk 6]
```diff
@@ -20,14 +21,29 @@ class SprykDumpConsole extends AbstractSprykConsole
 
     public const ARGUMENT_SPRYK = 'spryk';
 
+    public const OPTION_LEVEL = 'level';
+    public const OPTION_LEVEL_SHORT = 'l';
+
+    /**
+     * @var int|null
+     */
+    protected $level = SprykConfig::SPRYK_LEVEL_1;
```
[the remark this answers is not in the document]
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Console/SprykDumpConsole.php]: fixed

File: src/SprykerSdk/Spryk/Console/SprykDumpConsole.php [hunk 7]
```diff
@@ -20,14 +21,29 @@ class SprykDumpConsole extends AbstractSprykConsole
 
     public const ARGUMENT_SPRYK = 'spryk';
 
+    public const OPTION_LEVEL = 'level';
+    public const OPTION_LEVEL_SHORT = 'l';
```
[the remark this answers is not in the document]
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Console/SprykDumpConsole.php]: changed

File: src/SprykerSdk/Spryk/Console/SprykDumpConsole.php [hunk 8]
```diff
@@ -38,6 +54,9 @@ protected function configure()
      */
     protected function execute(InputInterface $input, OutputInterface $output): int
     {
+        $level = $input->getOption(self::OPTION_LEVEL);
```
[the remark this answers is not in the document]
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Console/SprykDumpConsole.php]: changed

File: src/SprykerSdk/Spryk/Console/SprykDumpConsole.php [hunk 9]
```diff
@@ -38,6 +54,9 @@ protected function configure()
...
+        $this->level = $level === 'all' ? null : (int) $level;
```
[the remark this answers is not in the document]
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Console/SprykDumpConsole.php]: moved

File: src/SprykerSdk/Spryk/Model/Spryk/Configuration/Loader/SprykConfigurationLoader.php [hunk 10]
```diff
@@ -37,29 +38,29 @@ class SprykConfigurationLoader implements SprykConfigurationLoaderInterface
     protected $configurationValidator;
 
     /**
-     * @var string
+     * @var SprykConfig
```
[the remark this answers is not in the document]
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Model/Spryk/Configuration/Loader/SprykConfigurationLoader.php]: fixed

File: src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php (hunk 1 above)
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php:14]: done

File: src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php (hunk 2 above)
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php]: fixed

File: src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php (hunk 3 above)
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Model/Spryk/Configuration/Validator/Rules/LevelRule.php:27]: added

File: src/SprykerSdk/Spryk/SprykConfig.php (hunk 4 above)
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/SprykConfig.php]: clarified the name of the method

File: src/SprykerSdk/Spryk/Console/SprykDumpConsole.php [hunk 11]
```diff
@@ -60,7 +79,7 @@ protected function dumpAllSpryks(OutputInterface $output): void
         $sprykDefinitions = $this->getFacade()->getSprykDefinitions();
```
[the remark this answers is not in the document]
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Console/SprykDumpConsole.php]: moved to facade level filtering

File: src/SprykerSdk/Spryk/SprykConfig.php [hunk 12]
```diff
@@ -25,6 +25,10 @@ class SprykConfig
     protected const NAME_ORGANIZATION = 'spryker-sdk';
     protected const NAME_PACKAGE = 'spryk';
 
+    public const SPRYK_LEVEL_1 = 1;
+    public const SPRYK_LEVEL_2 = 2;
+    public const SPRYK_LEVEL_3 = 3;
```
Comment (pushokwhite) [inline src/SprykerSdk/Spryk/SprykConfig.php:30]: It seems that we're using only first one as public constant, what if we move them all into protected part, but instead create a new public one like `public const SPRYK_DEFAULT_DUMP_LEVEL = self::SPRYK_LEVEL_1;`. (not a change request, just thinking)

File: src/SprykerSdk/Spryk/Model/Spryk/Dumper/SprykDefinitionDumper.php [hunk 13]
```diff
@@ -33,16 +33,20 @@ public function __construct(SprykDefinitionFinderInterface $definitionFinder, Sp
...
-    public function dump(): array
+    public function dump(?int $level = null): array
     {
         $sprykDefinitions = [];
         foreach ($this->definitionFinder->find() as $fileInfo) {
             $sprykName = str_replace('.' . $fileInfo->getExtension(), '', $fileInfo->getFilename());
             $sprykDefinition = $this->configurationLoader->loadSpryk($sprykName);
 
-            $sprykDefinitions[$sprykName] = $sprykDefinition;
+            if ($level === null || $level === (int) $sprykDefinition['level']) {
```
Comment (pushokwhite) [inline src/SprykerSdk/Spryk/Model/Spryk/Dumper/SprykDefinitionDumper.php:47]: It looks like a validation logic, could be moved to a separate method. But it's ok to leave it like this. Also please note that we're using the casting functions like `(int)` wihtout whitespaces: `(int)$smth`.

File: src/SprykerSdk/Spryk/SprykConfig.php [hunk 14]
```diff
@@ -138,6 +154,11 @@ public function getDefaultDevelopmentMode(): string
         return static::NAME_DEVELOPMENT_LAYER_PROJECT;
     }
 
+    public function getDefaultBuildLevel(): int
```
[the remark this answers is not in the document]
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/SprykConfig.php:162]: fixed

File: src/SprykerSdk/Spryk/Console/SprykDumpConsole.php [hunk 15]
```diff
@@ -38,29 +49,41 @@ protected function configure()
...
 
+    /**
+     * @param mixed $level
+     *
+     * @return int|null
+     */
+    protected function getLevelOption($level): ?int
+    {
+        return $level === 'all' ? null : (int) $level;
+    }
```
[the remark this answers is not in the document]
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/Console/SprykDumpConsole.php]: fixed

File: src/SprykerSdk/Spryk/SprykConfig.php (hunk 12 above)
Comment (oleksandr-stoilovskyi) [reply src/SprykerSdk/Spryk/SprykConfig.php]: fixed

**Review by pushokwhite — APPROVED** [review]

**Pull request merged**